File: python/499/D.py
class BIT:
    def __init__(self, n):
        self.n = n
        self.t = [0] * (n+1)
    def update(self, i, v):
        i += 1
        while i <= self.n:
            if v > self.t[i]: self.t[i] = v
            i += i & -i
    def query(self, i):
        i += 1
        r = 0
        while i > 0:
            if self.t[i] > r: r = self.t[i]
            i -= i & -i
        return r 


def maxAlternatingSum(self, nums: list[int], k: int) -> int:
    n = len(nums)
    vals = sorted(set(nums))
    comp = {v: i for i, v in enumerate(vals)}
    m = len(vals)
    bitLow = BIT(m)
    bitHigh = BIT(m)
    low = [0] * n
    high = [0] * n
    ans = 0
    for j in range(n):
        if j >= k:
            i = j-k
            ci = comp[nums[i]]
            bitLow.update(ci, low[i])
            bitHigh.update(m-1-ci, high[i]) 
        cj = comp[nums[j]]
        high[j] = nums[j] + (bitLow.query(cj-1) if cj > 0 else 0)
        rev = m - 1 - cj
        low[j] = nums[j] + (bitHigh.query(rev-1) if rev > 0 else 0)
        ans = max(ans, low[j], high[j])
    return ans

File: python/499/D.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cstdint>

struct BIT {
    int n;
    std::vector<int64_t> t;
    BIT(int _n) : n(_n), t(n+1,0) {}
    void update(int i, int64_t v) {
        for (i+=1; i <= n; i += i & -i)
            t[i] = std::max(t[i], v);
    }
    int64_t query(int i) {
        int64_t r = 0;
        for (i += 1; i > 0; i -= i & -i)
           r = std::max(r, t[i]);
        return r;
    }
};

long long maxAlternatingSum(std::vector<int>& nums, int k) {
    int n = nums.size();
    std::vector<int> vals = nums;
    std::sort(vals.begin(), vals.end());
    vals.erase(std::unique(vals.begin(), vals.end()), vals.end());
    auto rankOf = [&](int v) {
        return (int)(std::lower_bound(vals.begin(), vals.end(), v) - vals.begin());
    };
    int m= vals.size();
    BIT bitLow(m), bitHigh(m);
    std::vector<int64_t> low(n), high(n);
    int64_t ans = 0;
    for (int j =0; j < n; ++j) {
        if (j >= k) {
            int i = j-k;
            int ci = rankOf(nums[i]);
            bitLow.update(ci, low[i]);
            bitHigh.update(m-1-ci, high[i]);
        }
        int cj = rankOf(nums[j]);
        high[j] = nums[j] + (cj > 0 ? bitLow.query(cj-1) : 0LL);
        int rev = m-cj-1;
        low[j] = nums[j] + (rev > 0 ? bitHigh.query(rev-1) : 0LL);
        ans = std::max({ans, low[j], high[j]});
    }
    return ans;
}